core/src/mmu.cpp

```diff
@@ -2,54 +2,51 @@
 
 #include "mmu.h"
 
-#include <utility>
+#include <algorithm>
 
 namespace n_e_s::core {
 
-uint8_t Mmu::read_byte(uint16_t addr) const {
-    return *translate_address(addr);
-}
+Mmu::Mmu() : mem_banks_() {}
 
-uint16_t Mmu::read_word(uint16_t addr) const {
-    return read_byte(addr) | read_byte(addr + 1) << 8;
+void Mmu::add_mem_bank(IMemBank *mem_bank) {
+    mem_banks_.push_back(mem_bank);
 }
 
-void Mmu::write_byte(uint16_t addr, uint8_t byte) {
-    *translate_address(addr) = byte;
-}
+IMemBank *Mmu::get_mem_bank(uint16_t addr) const {
+    auto it = std::find_if(
+            begin(mem_banks_), end(mem_banks_), [=](IMemBank *mem_bank) {
+                return mem_bank->is_address_in_range(addr);
+            });
 
-void Mmu::write_word(uint16_t addr, uint16_t word) {
-    write_byte(addr, word & 0xFF);
-    write_byte(addr + 1, word >> 8);
+    if (it != end(mem_banks_)) {
+        return *it;
+    } else {
+        return nullptr;
+    }
 }
 
-uint8_t *Mmu::translate_address(uint16_t addr) {
-    return const_cast<uint8_t *>(std::as_const(*this).translate_address(addr));
+uint8_t Mmu::read_byte(uint16_t addr) const {
+    IMemBank *mem_bank = get_mem_bank(addr);
+    return mem_bank ? mem_bank->read_byte(addr) : 0;
 }
 
-const uint8_t *Mmu::translate_address(uint16_t addr) const {
-    if (addr <= ram_end_) {
-        addr %= ram_size_;
-        return &ram_[addr];
-    }
-
-    if (addr <= ppu_end_) {
-        addr %= ppu_size_;
-        return &ppu_[addr];
-    }
+uint16_t Mmu::read_word(uint16_t addr) const {
+    IMemBank *mem_bank = get_mem_bank(addr);
+    return mem_bank ? mem_bank->read_word(addr) : 0;
+}
 
-    if (addr <= io_end_) {
-        addr %= io_size_;
-        return &io_[addr];
+void Mmu::write_byte(uint16_t addr, uint8_t byte) {
```

Comment:
These should probably either assert that the address is fine or throw. Silently doing nothing seems difficult to debug.